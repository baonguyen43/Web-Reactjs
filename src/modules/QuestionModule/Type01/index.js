import React from 'react';
import classNames from 'classnames';

import { Button, Row, Col, ListGroup, ListGroupItem, Card, CardBody, CardImg, CardFooter } from 'reactstrap';
import Listen from 'components/Listening';
import { DictionaryText } from 'components/Dictionary';
import NotData from 'components/Error/NotData';
import PropTypes from 'prop-types';
import openNotificationWithIcon from 'components/Notification';

// const color = ['warning', 'success', 'danger', 'info', 'default'];
// const randomIndex = Math.floor(Math.random(color) * 5);
const name = 'bg-default justify-content-md-center mt-4';

class Type01 extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isDone: false,
      question: [],
      exerciseCountdowns: []
    };
    this.score = 0;
  }

  static getDerivedStateFromProps = (props, state) => {
    if (props.question !== state.question) {
      return {
        isDone: false,
        question: props.question,
      }
    }
    return null;
  }

  onListened = () => {
    this.setState({ isDone: true });
  }

  onNext = () => {
    this.checkAnswer();
  }

  checkAnswer = () => {
    const { onNext, questionIndex } = this.props;
    const { exerciseCountdowns } = this.state;
    const { isDone } = this.state;
    const isPush = exerciseCountdowns.findIndex((x) => x.questionIndex === questionIndex) > -1;
    if (!isPush) {
      if (isDone) {
        this.score = 100;
        openNotificationWithIcon('success', 'Notification', 'Correct');
      } else {
        this.score = 0;
        openNotificationWithIcon('danger', 'Notification', 'Incorrect');

      }
      exerciseCountdowns.push({ isDone, questionIndex });

      this.setState({ exerciseCountdowns })
    }
    const postAnswerToApiParams = {
      answerType: 'NEWWORD',
      questionGuid: '',
      notes: '',
      studentChoice: JSON.stringify({
        score: this.score
      }),
    }
    //Gọi props onNext tại questionIndex
    onNext(exerciseCountdowns, postAnswerToApiParams, isPush)
  }

  renderQuestion = (question) => {
    const { disabledBack, disabledNext, onBack } = this.props;
    return (
      <React.Fragment>
        <CardImg
          style={{ height: 300 }}
          alt="..."
          src={question.imageUrl}
        />
        <CardBody className='text-center'>
          <Listen
            audioURL={question.soundUrl}
            onListened={this.onListened}
            onAuto={false}
          />
          <p className={classNames(['question-type__word'])}>
            {question.text}
          </p>
          {question.phonetic !== '' ?
            <p className={classNames(['question-type__kindOfWord'])}>
              <i>{`/${question.phonetic}/`}</i>
            </p>
            : <span></span>
          }
          <p className={classNames(['question-type__kindOfWord'])}>
            <i>{`(${question.wordType})`}</i>
          </p>
          <p className={classNames(['question-type__kindOfWord'])}>
            {question.text_VN}
          </p>
        </CardBody>
        <CardFooter className="d-flex justify-content-between">
          <Button
            disabled={disabledBack}
            onClick={onBack}
          >Back
          </Button>
          <Button
            disabled={disabledNext}
            onClick={() => this.onNext()}
          >Next
          </Button>
        </CardFooter>
      </React.Fragment>
    )
  }

  renderAudio = (question) => {
    if (!question.examplesFormat) return null;
    return JSON.parse(question?.examplesFormat).sentences.map((item, index) => {
      return (
        <ListGroupItem className='text-center' key={index}>
          <strong className='text-center ml-2'>
            <DictionaryText text={item.text} />
          </strong>
          <div>
            <Listen
              custom
              audioURL={item.soundUrl}
              className={'question-type__audioExampleType02'}
            >
              <i className="fas fa-volume-up"></i>
            </Listen>
          </div>
        </ListGroupItem>
      )
    })
  }

  render() {
    const { question } = this.props;

    if (!question) {
      return <NotData />;
    }

    return (
      <div>
        {question.examplesFormat !== null ?
          <Row className={name}>
            <Col className='my-2 col-12 col-lg-6 text-center'>
              <Card style={{ height: '100%' }}>
                {this.renderQuestion(question)}
              </Card>
            </Col>
            <Col className="col-12 col-lg-6 text-center my-2">
              <Card style={{ height: '100%' }}>
                <CardBody>
                  <h4 className='text-warning'>Examples</h4>
                  <ListGroup>
                    {this.renderAudio(question)}
                  </ListGroup>
                </CardBody>
              </Card>
            </Col>
          </Row>
          :
          <Row>
            <Col className='mt-4'>
              <Card>{this.renderQuestion(question)}</Card>
            </Col>
          </Row>
        }
      </div>
    );
  }
}

Type01.propTypes = {
  // renderExerciseCountdowns: PropTypes.func.isRequired,
  onNext: PropTypes.func.isRequired,
  onBack: PropTypes.func.isRequired,
  questionIndex: PropTypes.number.isRequired,
  disabledNext: PropTypes.bool.isRequired,
  disabledBack: PropTypes.bool.isRequired,
  question: PropTypes.instanceOf(Object).isRequired,
}

export default Type01
