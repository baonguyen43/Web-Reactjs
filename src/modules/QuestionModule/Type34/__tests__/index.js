import React from "react";
import renderer from "react-test-renderer";

import Type01 from "../index";

test("type 01 module in ames247", () => {
  const component = renderer.create(
    <Type01
      word={""}
      imageURL={""}
      audioURL={""}
      kindOfWord={""}
      onNext={() => {}}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
