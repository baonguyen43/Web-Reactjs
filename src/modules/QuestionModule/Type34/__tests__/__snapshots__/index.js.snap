// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type 01 module in ames247 1`] = `
<div>
  <img
    className="question-type__image"
    src=""
  />
  <h1
    className="question-type__title"
  >
    NEW WORK
  </h1>
  <div
    className="ant-row-flex ant-row-flex-center ant-row-flex-top"
    style={undefined}
  >
    <div
      className="text-align_center"
    >
      <button
        className="btn-dashed btn-dashed__audio animated"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-volume-up fa-w-18 fa-4x "
          color="#EE6F41"
          data-icon="volume-up"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M215.03 71.05L126.06 160H24c-13.26 0-24 10.74-24 24v144c0 13.25 10.74 24 24 24h102.06l88.97 88.95c15.03 15.03 40.97 4.47 40.97-16.97V88.02c0-21.46-25.96-31.98-40.97-16.97zm233.32-51.08c-11.17-7.33-26.18-4.24-33.51 6.95-7.34 11.17-4.22 26.18 6.95 33.51 66.27 43.49 105.82 116.6 105.82 195.58 0 78.98-39.55 152.09-105.82 195.58-11.17 7.32-14.29 22.34-6.95 33.5 7.04 10.71 21.93 14.56 33.51 6.95C528.27 439.58 576 351.33 576 256S528.27 72.43 448.35 19.97zM480 256c0-63.53-32.06-121.94-85.77-156.24-11.19-7.14-26.03-3.82-33.12 7.46s-3.78 26.21 7.41 33.36C408.27 165.97 432 209.11 432 256s-23.73 90.03-63.48 115.42c-11.19 7.14-14.5 22.07-7.41 33.36 6.51 10.36 21.12 15.14 33.12 7.46C447.94 377.94 480 319.54 480 256zm-141.77-76.87c-11.58-6.33-26.19-2.16-32.61 9.45-6.39 11.61-2.16 26.2 9.45 32.61C327.98 228.28 336 241.63 336 256c0 14.38-8.02 27.72-20.92 34.81-11.61 6.41-15.84 21-9.45 32.61 6.43 11.66 21.05 15.8 32.61 9.45 28.23-15.55 45.77-45 45.77-76.88s-17.54-61.32-45.78-76.86z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="btn-dashed btn-dashed__reset"
        onBlur={undefined}
        onClick={[Function]}
        onContextMenu={undefined}
        onFocus={undefined}
        onMouseDown={undefined}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={undefined}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-undo fa-w-16 fa-1x "
          color="#EE6F41"
          data-icon="undo"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <p
    className="question-type__word"
  >
    
  </p>
  <p
    className="question-type__kindOfWord"
  >
    (  )
  </p>
  <p
    className="question-type__note"
  >
    Click the icon above to make sure you can hear the tone clearly.
  </p>
  <button
    className="ant-btn question-type__button ant-btn-primary ant-btn-lg"
    onClick={[Function]}
    title={undefined}
    type="button"
  >
    <i
      className="anticon anticon-check"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="check"
        fill="currentColor"
        height="1em"
        onClick={undefined}
        style={undefined}
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 0 0-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
        />
      </svg>
    </i>
    <span>
      Next question
    </span>
  </button>
</div>
`;

exports[`type 01 module in ames247 2`] = `
<div>
  <img
    className="question-type__image"
    src=""
  />
  <h1
    className="question-type__title"
  >
    NEW WORK
  </h1>
  <div
    className="ant-row-flex ant-row-flex-center ant-row-flex-top"
    style={undefined}
  >
    <div
      className="text-align_center"
    >
      <button
        className="btn-dashed btn-dashed__audio animated"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-volume-up fa-w-18 fa-4x "
          color="#EE6F41"
          data-icon="volume-up"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M215.03 71.05L126.06 160H24c-13.26 0-24 10.74-24 24v144c0 13.25 10.74 24 24 24h102.06l88.97 88.95c15.03 15.03 40.97 4.47 40.97-16.97V88.02c0-21.46-25.96-31.98-40.97-16.97zm233.32-51.08c-11.17-7.33-26.18-4.24-33.51 6.95-7.34 11.17-4.22 26.18 6.95 33.51 66.27 43.49 105.82 116.6 105.82 195.58 0 78.98-39.55 152.09-105.82 195.58-11.17 7.32-14.29 22.34-6.95 33.5 7.04 10.71 21.93 14.56 33.51 6.95C528.27 439.58 576 351.33 576 256S528.27 72.43 448.35 19.97zM480 256c0-63.53-32.06-121.94-85.77-156.24-11.19-7.14-26.03-3.82-33.12 7.46s-3.78 26.21 7.41 33.36C408.27 165.97 432 209.11 432 256s-23.73 90.03-63.48 115.42c-11.19 7.14-14.5 22.07-7.41 33.36 6.51 10.36 21.12 15.14 33.12 7.46C447.94 377.94 480 319.54 480 256zm-141.77-76.87c-11.58-6.33-26.19-2.16-32.61 9.45-6.39 11.61-2.16 26.2 9.45 32.61C327.98 228.28 336 241.63 336 256c0 14.38-8.02 27.72-20.92 34.81-11.61 6.41-15.84 21-9.45 32.61 6.43 11.66 21.05 15.8 32.61 9.45 28.23-15.55 45.77-45 45.77-76.88s-17.54-61.32-45.78-76.86z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="btn-dashed btn-dashed__reset"
        onBlur={undefined}
        onClick={[Function]}
        onContextMenu={undefined}
        onFocus={undefined}
        onMouseDown={undefined}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={undefined}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-undo fa-w-16 fa-1x "
          color="#EE6F41"
          data-icon="undo"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <p
    className="question-type__word"
  >
    
  </p>
  <p
    className="question-type__kindOfWord"
  >
    (  )
  </p>
  <p
    className="question-type__note"
  >
    Click the icon above to make sure you can hear the tone clearly.
  </p>
  <button
    className="ant-btn question-type__button ant-btn-primary ant-btn-lg"
    onClick={[Function]}
    title={undefined}
    type="button"
  >
    <i
      className="anticon anticon-check"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="check"
        fill="currentColor"
        height="1em"
        onClick={undefined}
        style={undefined}
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 0 0-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
        />
      </svg>
    </i>
    <span>
      Next question
    </span>
  </button>
</div>
`;
